### PR TITLE
DFlash2 Support

### PR DESCRIPTION
**Pull request by hikettei**
```sh
$  uv run lalamo speculator convert z-lab/Qwen3.8-27B-DFlash2 --output-dir <somewhere>
```

### CODE REVIEW

**Comment by CC-Yeh** [comment]
@hikettei 
Just to confirm: we don't use `CandidateSelector` from DFlash2 right?

File: tests/unit/test_model_import_loading.py [hunk 1]
```diff
@@ -604,3 +648,37 @@ def test_model_export_load_with_strong_initializer_forces_saved_float_dtypes(tmp
...
+    kernel_projection = jnp.arange(32, dtype=jnp.float32).reshape(8, 4)
+    predecessor_codebook = jnp.arange(32, dtype=jnp.float32).reshape(16, 2)
+    successor_codebook = predecessor_codebook + 1
+    hidden_projection = jnp.arange(8, dtype=jnp.float32).reshape(2, 4)
+    weights = {
+        "attention_conv.base_kernel": base_kernel,
+        "attention_conv.kernel_projection.weight": kernel_projection,
+        "candidate_selector.predecessor_codebook": predecessor_codebook,
+        "candidate_selector.successor_codebook": successor_codebook,
+        "candidate_selector.hidden_projection.weight": hidden_projection,
```
Comment (CC-Yeh) [inline tests/unit/test_model_import_loading.py]: will this export unused `candidate_selector` weights?

Comment (hikettei) [reply tests/unit/test_model_import_loading.py]: if you export the hf one yes, but my training checkpoint drops candidate_selector basically

**Comment by CC-Yeh** [comment]
CandidateSelector appears to be instantiated, loaded, and exported, but never participates in proposal/runtime execution.

`DFlashDraftConfig.init()` creates it (`lalamo/modules/speculators/dflash.py`), and `load_dflash_draft_model()` loads it via `load_candidate_selector()` (`lalamo/model_import/loaders/dflash_loader.py`). However, `DFlashDraftModel.__call__()` only applies the grouped-convolution transforms, layers, and output norm; it never reads `self.candidate_selector`.

In the generated PR356 float32 artifact, this produces these otherwise-unused tensors:
```
speculator.draft_model.candidate_selector.predecessor_codebook
speculator.draft_model.candidate_selector.successor_codebook
speculator.draft_model.candidate_selector.hidden_projection.weights.weights
```
Quantized exports may add further CandidateSelector arrays. Strict consumers that validate every safetensors tensor must either carry dead parameters or add special ignore logic. The current test verifies loading (`test_load_dflash2_components`), but not proposal behavior.

Could we either wire CandidateSelector into candidate/proposal generation and add an integration test, or omit it from the export until semantics are implemented? Making export explicitly optional would also be reasonable.

File: tests/unit/test_model_import_loading.py (hunk 1 above)
Comment (hikettei) [reply tests/unit/test_model_import_loading.py]: removed

File: lalamo/model_import/model_configs/huggingface/dflash.py [hunk 2]
```diff
@@ -167,6 +189,9 @@ def to_dflash_draft_config(self) -> DFlashDraftConfig:
             up_clipping=None,
         )
         rope_config = self._rope_config(self.max_position_embeddings)
+        grouped_convolution_config = self.grouped_convolution_config(linear_config)
+        if architecture == "DFlash2DraftModel" and grouped_convolution_config is None:
+            raise ValueError("DFlash2DraftModel requires a grouped convolution config.")
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/dflash.py:194]: this doesn't look very compatible with uzu tbh

File: lalamo/model_import/model_configs/huggingface/dflash.py [hunk 3]
```diff
@@ -147,7 +153,23 @@ def _layer_sliding_window_sizes(self) -> tuple[int | None, ...]:
             self.sliding_window if layer_type == "sliding_attention" else None for layer_type in self.layer_types
         )
 
+    def grouped_convolution_config(self, linear_config: LinearConfig) -> DFlashGroupedConvolutionConfig | None:
+        kernel_size = self.dflash_config.conv_kernel_size
+        group_size = self.dflash_config.conv_group_size
+        if kernel_size is None and group_size is None:
+            return None
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/dflash.py:160]: useless condition?

**Review by knyazer — CHANGES_REQUESTED** [review]
also i agree with artur that convs should probably be represented via our convs module

**Comment by CC-Yeh** [comment]
CC @knyazer — we received inference-side architecture feedback that the current TransformerSublayerTransform approach models DFlash v2 as an injected wrapper rather than as actual layer topology.

Proposed cross-repo alignment:

- TransformerLayer directly owns optional attention_convolution and mlp_convolution fields.
- TransformerLayer performs prepare → attention/MLP → finish itself at the norm/residual boundaries.
- Remove TransformerSublayerTransform, mixer_transform/mlp_transform call arguments, DFlashLayerGroupedConvolutions, and DFlashDraftModel.layer_grouped_convolutions.
- HFDFlashConfig should lower the shared HF conv_kernel_size/conv_group_size values into the transformer-layer construction/configuration, while each layer still creates two independent parameter modules.
- Exported weights should live with their layer-local attention/MLP convolution modules.

The convolution needs only the normalized sublayer input and final sublayer output, not attention or MLP internals, so TransformerLayer appears to be the correct owner. Does this match the Lalamo direction you would prefer? We want both repositories to expose the same topology and avoid independently preserving the transform seam.

**Review by norpadon — APPROVED** [review]

File: lalamo/modules/transformer_layer.py [hunk 4]
```diff
@@ -167,6 +169,34 @@ class TransformerLayerConfig(LalamoConfig):
     has_post_layer_scalar: bool = False
     kv_source_layer_index: int | None = None
     rope_config: RoPEConfig | None = None
+    conv_config: SeparableCausalConvConfig | None = None
+    kernel_projection_config: LinearConfig | None = None
+    conv_kernel_size: int | None = None
+    conv_group_size: int | None = None
```
Comment (uuuvn) [inline lalamo/modules/transformer_layer.py]: 1). mixer pre/post duo and mlp pre/post duo are logcially separate things, let's make them separate instead of one config that controls both and then init creates either two sublayer transforms or no sublayer transforms

2). Why is there 4 fields which are all individual `Something | None` but are always asserted either all `Something` or if `conv_config` is `None` three other fields are just ignored? I would've done `(mixer/mlp)_conv_config: TransformerLayerConvConfig | None` for this

3). (preexisting with `kernel_size`, added same style `group_size`) Why are `conv_kernel_size`/`conv_group_size` not part of `SeparableCausalConvConfig` like `has_biases` already is?

Comment (uuuvn) [inline lalamo/modules/transformer_layer.py:175]: 1). mixer pre/post duo and mlp pre/post duo are logcially separate things, let's make them separate instead of one config that controls both and then init creates either two sublayer transforms or no sublayer transforms

2). Why is there 4 fields which are all individual `Something | None` but are always asserted either all `Something` or if `conv_config` is `None` three other fields are just ignored? I would've done `(mixer/mlp)_conv_config: TransformerLayerConvConfig | None` for this

3). (preexisting with `kernel_size`, added same style `group_size`) Why are `conv_kernel_size`/`conv_group_size` not part of `SeparableCausalConvConfig` like `has_biases` already is?

File: lalamo/modules/transformer_layer.py [hunk 5]
```diff
@@ -167,6 +169,34 @@ class TransformerLayerConfig(LalamoConfig):
...
+
+    def init_sublayer_transform(
+        self,
+        initializer: Initializer,
+        model_dim: int,
+    ) -> tuple[SeparableCausalConv, SeparableCausalConv, Linear] | tuple[None, None, None]:
```
Comment (uuuvn) [inline lalamo/modules/transformer_layer.py:181]: Big tuple that's either three `Something` or three `None`, I think it makes sense to have a named dataclass like `TransformerLayerConv` for those three things and have `-> TransformerLayerConv | None` in return type here.

File: lalamo/modules/transformer_layer.py [hunk 6]
```diff
@@ -209,6 +247,56 @@ class TransformerLayer(LalamoModule[TransformerLayerConfig]):
     post_mlp_norm: Normalization | None
     ple: PLELayer | None
     post_layer_scalar: Float[Array, "1"] | None
+    pre_mixer_conv: SeparableCausalConv | None
+    post_mixer_conv: SeparableCausalConv | None
+    mixer_kernel_projection: Linear | None
+    pre_mlp_conv: SeparableCausalConv | None
+    post_mlp_conv: SeparableCausalConv | None
+    mlp_kernel_projection: Linear | None
```
Comment (uuuvn) [inline lalamo/modules/transformer_layer.py:255]: Why are all 6 at the end? I think it makes sense to have them be in a dataclass in the same order data flows through them.

**Review by uuuvn — CHANGES_REQUESTED** [review]
Left a couple comments, also please provide a migration script to migrate old models to the new format

File: lalamo/modules/transformer_layer.py (hunk 6 above)
Comment (hikettei) [reply lalamo/modules/transformer_layer.py]: hmph, okay lets revert to the old implementation

Comment (norpadon) [reply lalamo/modules/transformer_layer.py]: Please no. Lets leave it like this for now and wait for a new layer implementation

File: lalamo/modules/transformer_layer.py (hunk 4 above)
Comment (hikettei) [reply lalamo/modules/transformer_layer.py]: > (preexisting with kernel_size, added same style group_size) Why are conv_kernel_size/conv_group_size not part of SeparableCausalConvConfig like has_biases already is?

To follow the existing `Mamba2Config` style (idk about preexisting config design), but i dont want to introduce exception here. We can refactor them together later.
